File: Lectures/Lecture-7/BFS.py
N = 20
visited = [False]*N
dist = [0]*N
adj = [[] for i in range(N)]

def BFS(source):
	queue = []
	idx = 0
	queue.append(source)
	visited[source] = True
	
	while idx < len(queue): # This is the same as while(!q.empty())
		v = queue[idx]
		idx += 1
		for w in adj[v]:
			if visited[w]:
				continue
			queue.append(w)
			visited[w]=True
			dist[w] = dist[v] + 1

with open("BFS_input.txt", 'r') as in_file:
	line_number=0
	n,e = 0,0
	for line in in_file:
		if line_number == 0:
			n,m=[int(x) for x in line.split()]
		else:
			a,b=[int(x) for x in line.split()]
			adj[a].append(b)
			adj[b].append(a)
		line_number += 1

s = 1
BFS(s)

for i in range(1,n+1):
	print(f"Distance between source (1) and {i} = {dist[i]}")

'''
[1, 2, 5, 3, 4, 6]
                   ^

'''

File: Lectures/Lecture-7/BFS.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

const int N = 20;
vector<int> adj[N];

int n, e;
int dist[N];
bool visited[N];

void BFS(int u) {
	for (int i = 0; i <= n; i++) {
		dist[i] = 0;
		visited[i] = false;
	}

	queue<int> q;
	q.push(u); visited[u] = true;
	while (!q.empty()) {
		int v = q.front();
		q.pop();

		for (int w : adj[v]) {
			if (visited[w])
				continue;
			q.push(w); visited[w] = true;
			dist[w] = dist[v] + 1;
		}
	}
	return;
}


int main() {
	FILE* in_file, * out_file;
	freopen_s(&in_file, "input.txt", "r", stdin);
	freopen_s(&out_file, "output.txt", "w", stdout);


	cin >> n >> e;
	for (int i = 0; i < e; i++) {
		int u, v;
		cin >> u >> v;
		adj[u].push_back(v);
		adj[v].push_back(u);
	}

	int s = 1;
	BFS(s);

	for (int i = 1; i <= n; i++) {
		cout << "Distance between source (1) and " << i << " = " << dist[i] << endl;
	}
}
